codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -1230,210 +876,114 @@ public String toString() {
+            iter = map.entrySet().iterator();
         }
 
         @Override
-        public T getKey() {
-            return name;
+        public boolean hasNext() {
+            return iter.hasNext() || values != null;
         }
 
         @Override
-        public T getValue() {
-            return value;
+        @SuppressWarnings("unchecked")
+        public Entry<T, T> next() {
+            if (name == null) {
+                Entry<T, Object> entry = iter.next();
+                if (!isList(entry)) {
+                    return (Entry<T, T>) entry;
+                }
+                initListIterator(entry);
+            }
+            return fetchNextEntryFromList();
         }
 
-        @Override
-        public T setValue(T value) {
-            checkNotNull(value, "value");
-            T oldValue = this.value;
-            this.value = value;
-            return oldValue;
+        @SuppressWarnings("unchecked")
+        private void initListIterator(Entry<T, Object> entry) {
+            name = entry.getKey();
+            values = (List<T>) entry.getValue();
+            valuesIdx = 0;
+        }
+
+        private Entry<T, T> fetchNextEntryFromList() {
+            Entry<T, T> next = new SimpleListEntry<T>(name, values, valuesIdx);
+            valuesIdx++;
+            if (values.size() == valuesIdx) {
+                name = null;
+                values = null;
+            }
+            return next;
         }
 
         @Override
-        public String toString() {
-            return new StringBuilder()
-                .append(name)
-                .append('=')
-                .append(value)
-                .toString();
+        public void remove() {
+            throw new UnsupportedOperationException();
         }
     }
 
-    protected final class KeyValueHeaderIterator implements Iterator<Entry<T, T>> {
+    private static class SimpleListEntry<T> extends SimpleEntry<T, T> {
+        private static final long serialVersionUID = 542062057061955051L;
+        private final List<T> values;
+        private final int idx;
 
-        private HeaderEntry current = head;
-
-        @Override
-        public boolean hasNext() {
-            return current.after != head;
+        SimpleListEntry(T name, List<T> values, int idx) {
+            super(name, values.get(idx));
+            this.values = values;
+            this.idx = idx;
         }
 
         @Override
-        public Entry<T, T> next() {
-            current = current.after;
-
-            if (current == head) {
-                throw new NoSuchElementException();
-            }
-
-            return current;
+        public T setValue(T value) {
+            T oldValue = super.setValue(value);
+            values.set(idx, value);
+            return oldValue;
         }
+    }
 
-        @Override
-        public void remove() {
-            throw new UnsupportedOperationException();
+    private static class ValuesList<T> extends ArrayList<T> {
```

Comment:
This was to address @Scottmitch's concern about not being able to have values that are ArrayLists:
> minor nit: we are checking isList(..) and assuming this means we have accounted for multiple values, but what if the generic type is a List.

This class has effectively zero overhead over ArrayList itself, so it seemed reasonable to just remove the restriction.